### PR TITLE
4 of 5 read methods for permission state.

### PR DESCRIPTION
**Pull request by hmlanigan**
Implementation of 4 of 5 Read methods in the domain/permission/state package.

- ReadUserAccessForTarget
- ReadUserAccessLevelForTarget
- ReadAllUserAccessForUser
- ReadAllUserAccessForTarget

ReadAllAccessTypeForUser will be implemented in the next PR.

This PR will need to be rebased once #17067 lands.

## Checklist

- [x] Code style: imports ordered, good names, simple structure, etc
- [x] Comments saying why design decisions were made
- [x] Go unit tests, with comments saying what you're testing
- [ ] ~[Integration tests](https://github.com/juju/juju/tree/main/tests), with comments saying what you're testing~
- [ ] ~[doc.go](https://discourse.charmhub.io/t/readme-in-packages/451) added or updated in changed packages~

## QA steps

These methods are not used yet, tested via unit tests.

## Links

**Jira card:** JUJU-5558

### CODE REVIEW

File: domain/permission/state/state.go [hunk 1]
```diff
@@ -195,25 +196,153 @@ func (s *State) UpsertPermission(ctx context.Context, args permission.UpsertPerm
...
+		}
+		userAccess.Access = corepermission.Access(accessType)
+		userAccess.PermissionID = permissionUUID
+		userAccess.Object = objectTag(target)
+		return nil
+	})
+	if err == nil {
+		return userAccess, nil
+	}
+	return corepermission.UserAccess{}, errors.Trace(domain.CoerceError(err))
```
Comment (SimonRichardson) [inline domain/permission/state/state.go:226]: nit: flip this, do the error checking first.

File: domain/permission/state/state.go [hunk 2]
```diff
@@ -195,25 +196,153 @@ func (s *State) UpsertPermission(ctx context.Context, args permission.UpsertPerm
...
+		if err != nil {
+			return errors.Trace(err)
+		}
+		userAccessType = corepermission.Access(accessType)
+		return nil
+	})
+	if err == nil {
+		return userAccessType, nil
+	}
+	return userAccessType, errors.Trace(domain.CoerceError(err))
```
Comment (SimonRichardson) [inline domain/permission/state/state.go:256]: Same as above.

File: domain/permission/state/state.go [hunk 3]
```diff
@@ -195,25 +196,153 @@ func (s *State) UpsertPermission(ctx context.Context, args permission.UpsertPerm
...
+
+		permissions, err = readUsersPermissions(ctx, tx, user.UUID)
+		if err != nil {
+			return errors.Trace(err)
+		}
+
+		permissions, err = grantOnType(ctx, tx, permissions)
+		if err != nil {
+			return errors.Trace(err)
+		}
```
Comment (SimonRichardson) [inline domain/permission/state/state.go:284]: Let's fill in the outer value when we have all the data.

```suggestion
		userPermissions, err := readUsersPermissions(ctx, tx, user.UUID)
		if err != nil {
			return errors.Trace(err)
		}

		permissions, err = grantOnType(ctx, tx, userPermissions)
		if err != nil {
			return errors.Trace(err)
		}
```

File: domain/permission/state/state.go [hunk 4]
```diff
@@ -195,25 +196,153 @@ func (s *State) UpsertPermission(ctx context.Context, args permission.UpsertPerm
...
+
+		permissions, err = targetPermissions(ctx, tx, target.Key)
+		if err != nil {
+			return errors.Trace(err)
+		}
+
+		userUUIDs := make([]string, len(permissions))
+		for i, p := range permissions {
+			userUUIDs[i] = p.GrantTo
+		}
```
Comment (SimonRichardson) [inline domain/permission/state/state.go:323]: Shouldn't this be a map? That way you can de-duplicate reoccurring grants.

File: domain/permission/state/state.go [hunk 5]
```diff
@@ -355,21 +522,22 @@ SELECT &M.found_it FROM (
...
 	err = tx.Query(ctx, targetExistsStmt, sqlair.M{"grant_on": targetKey}).GetAll(&foundIt)
 	if err != nil {
-		return errors.Annotatef(err, "verifying %q target exists", targetKey)
+		return "", errors.Annotatef(err, "verifying %q target exists", targetKey)
 	}
 
 	if len(foundIt) == 1 {
-		return nil
+		return foundIt[0]["found_it"].(string), nil
 	}
```
Comment (SimonRichardson) [inline domain/permission/state/state.go:530]: Much better.

File: domain/permission/state/state.go [hunk 6]
```diff
@@ -384,3 +552,117 @@ func objectTag(id corepermission.ID) (result names.Tag) {
...
+	findAccessTypeStmt, err := sqlair.Prepare(findAccessTypeQuery, sqlair.M{})
+	if err != nil {
+		return "", "", errors.Annotate(err, "preparing select findAccessType query")
+	}
+
+	input := sqlair.M{
+		"grant_on": grantOn,
+		"grant_to": grantTo,
+	}
+	result := sqlair.M{}
```
Comment (SimonRichardson) [inline domain/permission/state/state.go:579]: This should be a type, then you don't have to type case on the return.

File: domain/permission/state/state.go (hunk 4 above)
Comment (hmlanigan) [reply domain/permission/state/state.go:323]: No, the permission table has a unique constraint that each combination of grant_on/grant_to. This method is returning all users (grant_to) who have a permission on a single target (grant_on). 

As well, if we used a map, and there were multiple combinations of grant_on/grant_to we'd get a random result.

File: domain/permission/state/state.go (hunk 1 above)
Comment (hmlanigan) [reply domain/permission/state/state.go:226]: In many ways I dislike the pattern of:
```
if err != nil{
  return errors.Trace(err)
}
```
We're doing the `if err nil (or not nil)` check twice. Just return `errors.Trace(err)`. The first thing both errors.Trace and domain.CoerceError do is check err nil and return nil if it is.

Comment (hmlanigan) [reply domain/permission/state/state.go:226]: I'd rather
```
return userAccess, errors.Trace(domain.CoerceError(err))
```

Comment (SimonRichardson) [reply domain/permission/state/state.go:226]: Personally, I'd rather be consistent with everywhere else. Return early on errors.

File: domain/permission/state/state.go (hunk 4 above)
Comment (SimonRichardson) [reply domain/permission/state/state.go:323]: IN doesn't guarantee order in sqlite, you would need to add an order by to ensure that then.

**Review by SimonRichardson — APPROVED** [review]

File: domain/permission/state/state.go [hunk 7]
```diff
@@ -384,3 +551,117 @@ func objectTag(id corepermission.ID) (result names.Tag) {
...
+// permission for the given grantOn and grantTo combination.
+// There can be only one.
+func (st *State) findAccessType(
+	ctx context.Context,
+	tx *sqlair.TX,
+	grantOn, grantTo string,
+) (string, string, error) {
+	findAccessTypeQuery := `
+SELECT type AS &readUserPermission.access_type, permission.uuid AS &readUserPermission.uuid
+FROM permission_access_type
```
Comment (manadart) [inline domain/permission/state/state.go:565]: Alias these for readability.

File: domain/permission/state/state.go [hunk 8]
```diff
@@ -384,3 +551,117 @@ func objectTag(id corepermission.ID) (result names.Tag) {
...
+     JOIN permission ON permission_access_type.id = permission.permission_type_id
+WHERE permission.grant_on = $readUserPermission.grant_on AND permission.grant_to = $readUserPermission.grant_to
```
Comment (manadart) [inline domain/permission/state/state.go:567]: Vertically align these clauses.

File: domain/permission/state/state.go [hunk 9]
```diff
@@ -384,3 +551,117 @@ func objectTag(id corepermission.ID) (result names.Tag) {
...
+// grantOn.
+func (st *State) targetPermissions(ctx context.Context,
+	tx *sqlair.TX,
+	grantOn string,
+) ([]readUserPermission, error) {
+	query := `
+SELECT (permission.uuid, permission.grant_on, permission.grant_to) AS (&readUserPermission.*),
+       permission_access_type.type AS &readUserPermission.access_type
+FROM permission
+     JOIN permission_access_type
```
Comment (manadart) [inline domain/permission/state/state.go:647]: Third time warrants a view.

File: domain/permission/state/state.go [hunk 10]
```diff
@@ -188,59 +189,187 @@ WHERE grant_to = $M.grant_to AND grant_on = $M.grant_on
...
+		permissions []readUserPermission
+		user        user
+	)
+	err = db.Txn(ctx, func(ctx context.Context, tx *sqlair.TX) error {
+		user, err = st.findUserByName(ctx, tx, subject)
+		if err != nil {
+			return errors.Trace(err)
+		}
+
+		userPermissions, err := st.readUsersPermissions(ctx, tx, user.UUID)
```
Comment (manadart) [inline domain/permission/state/state.go:276]: You should be able to write a new query for this that joins to the union, so you don't have to potato in `grantOnType`.

File: domain/permission/state/state.go (hunk 9 above)
Comment (hmlanigan) [reply domain/permission/state/state.go:647]: @manadart The entire thing? The `JOIN permission_access_type` part? It's not obvious based on our current usage of view how to do the small part.  (And being newer to SQL stmts). More detail is appreciated.

File: domain/permission/state/state.go (hunk 10 above)
Comment (hmlanigan) [reply domain/permission/state/state.go:276]: @manadart Not sure what you mean by "potato" here. And I'll need help on how to combine the queries. I was happy to get what I have. :-)

**Pull request merged**